barretenberg/cpp/src/barretenberg/vm/avm_trace/avm_trace.cpp

```diff
@@ -4382,9 +4383,16 @@ std::vector<Row> AvmTraceBuilder::finalize(uint32_t min_trace_size, bool range_c
     // 4. Whenever we hit the last row, we zero all write_offsets such that the shift relation will succeed
     std::vector<AvmKernelTraceBuilder::KernelTraceEntry> kernel_trace = kernel_trace_builder.finalize();
     size_t kernel_padding_main_trace_bottom = 1;
-    for (size_t i = 0; i < kernel_trace.size(); i++) {
-        auto const& src = kernel_trace.at(i);
-        // check the clock and iterate through the main trace until we hit the clock
+
+    // Index 1 corresponds here to the first active row of the main execution trace, as
+    // we already prepended the extra row for shifted columns. Therefore, initialization
+    // of side_effect_counter occurs occurs on this row.
+    main_trace.at(1).avm_kernel_side_effect_counter = initial_side_effect_counter;
```

Comment:
I think I kind of understand it now. My only question is if it wouldnt be better to just do this when you do

```
// Adding extra row for the shifted values at the top of the execution trace.
    Row first_row = Row{ .avm_main_first = FF(1), .avm_mem_lastAccess = FF(1) };
```

i.e., adding `.avm_kernel_side_effect_counter = initial_side_effect_counter`.